src/main/scala/io/iohk/ethereum/jsonrpc/FilterManager.scala

```diff
@@ -0,0 +1,253 @@
+        .map { case (log, logIndex) =>
+          val blockBody = blockchain.getBlockBodyByHash(blockHeader.hash).get
+          val tx = blockBody.transactionList(txIndex)
+          Log(
+            logIndex = logIndex,
+            transactionIndex = Some(txIndex), // TODO: None is this is a pending block (?)
+            transactionHash = Some(tx.hash), // TODO: None is this is a pending block (?)
+            blockHash = blockHeader.hash,
+            blockNumber = blockHeader.number,
+            address = log.loggerAddress,
+            data = log.data,
+            topics = log.logTopics)
+        }
+      } else logsSoFar
+    }
+  }
+
+  private def topicsMatch(logTopics: Seq[ByteString], filterTopics: Seq[Seq[ByteString]]): Boolean = {
+    (filterTopics zip logTopics).forall { case (filter, log) => filter.isEmpty || filter.contains(log) }
+  }
+
+  private def getBlockHashesAfter(blockNumber: BigInt): Seq[ByteString] = {
+    val bestBlock = appStateStorage.getBestBlockNumber()
+
+    @tailrec
+    def recur(currentBlockNumber: BigInt, hashesSoFar: Seq[ByteString]): Seq[ByteString] = {
+      if (currentBlockNumber > bestBlock) {
+        hashesSoFar
+      } else blockchain.getBlockHeaderByNumber(currentBlockNumber) match {
+        case Some(header) => recur(currentBlockNumber + 1, hashesSoFar :+ header.hash)
+        case None => hashesSoFar
+      }
+    }
+
+    recur(blockNumber + 1, Nil)
+  }
+
+  private def getPendingTransactions(): Future[Seq[SignedTransaction]] = {
+    (pendingTransactionsManager ? PendingTransactionsManager.GetPendingTransactions)
+      .mapTo[PendingTransactionsManager.PendingTransactions]
+      .flatMap { case PendingTransactionsManager.PendingTransactions(pendingTransactions) =>
+        keyStore.listAccounts() match {
+          case Right(accounts) => Future.successful(pendingTransactions.filter(pt => accounts.contains(pt.senderAddress)))
+          case Left(_) => Future.failed(new RuntimeException("Cannot get account list"))
+        }
+      }
+  }
+
+  private def getFilterById(id: BigInt): Option[Filter] =
+    filters.find(_.id == id)
+
+  private def generateId(): BigInt = Math.abs(Random.nextLong())
+
+  private def resolveBlockNumber(blockParam: BlockParam, appStateStorage: AppStateStorage): BigInt = {
+    blockParam match {
+      case BlockParam.WithNumber(blockNumber) => blockNumber
+      case BlockParam.Earliest => 0
+      case BlockParam.Latest => appStateStorage.getBestBlockNumber()
+      case BlockParam.Pending => appStateStorage.getBestBlockNumber() + 1
+    }
+  }
+}
+
+object FilterManager {
+  def props(blockchain: Blockchain,
+            appStateStorage: AppStateStorage,
+            keyStore: KeyStore,
+            pendingTransactionsManager: ActorRef): Props =
+    Props(new FilterManager(blockchain, appStateStorage, keyStore, pendingTransactionsManager))
+
+  sealed trait Filter {
+    def id: BigInt
+  }
+  case class LogFilter(
```

Comment:
Is it possible the following field is missing? 

`removed: TAG - true when the log was removed, due to a chain reorganization. false if its a valid log.`